File: tests/integration/aliases/go/retype_parents/step2/main.go
// Copyright 2016-2020, Pulumi Corporation.  All rights reserved.
//go:build !all
// +build !all

package main

import (
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type Resource struct {
	pulumi.ResourceState
}

type ComponentSix struct {
	pulumi.ResourceState
}

type ComponentSixParent struct {
	pulumi.ResourceState
}

func NewResource(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) (*Resource, error) {
	comp := &Resource{}
	err := ctx.RegisterComponentResource("my:module:Resource", name, comp, opts...)
	if err != nil {
		return nil, err
	}
	return comp, nil
}

// Scenario #6 - Nested parents changing types
func NewComponentSix(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) (*ComponentSix, error) {
	comp := &ComponentSix{}
	aliases := make([]pulumi.Alias, 0)
	for i := 0; i < 100; i++ {
		aliases = append(aliases, pulumi.Alias{
			Type: pulumi.Sprintf("my:module:ComponentSix-v%d", i),
		})
	}
	err := ctx.RegisterComponentResource(
		"my:module:ComponentSix-v0", name, comp,
		pulumi.Aliases(aliases), pulumi.Composite(opts...))
	if err != nil {
		return nil, err
	}
	parentOpt := pulumi.Parent(comp)
	_, err = NewResource(ctx, "otherchild", parentOpt)
	if err != nil {
		return nil, err
	}
	return comp, nil
}

func NewComponentSixParent(ctx *pulumi.Context, name string,
	opts ...pulumi.ResourceOption,
) (*ComponentSixParent, error) {
	comp := &ComponentSixParent{}
	aliases := make([]pulumi.Alias, 0)
	for i := 0; i < 10; i++ {
		aliases = append(aliases, pulumi.Alias{
			Type: pulumi.Sprintf("my:module:ComponentSixParent-v%d", i),
		})
	}
	err := ctx.RegisterComponentResource(
		"my:module:ComponentSixParent-v0", name, comp,
		pulumi.Aliases(aliases), pulumi.Composite(opts...))
	if err != nil {
		return nil, err
	}
	parentOpt := pulumi.Parent(comp)
	_, err = NewComponentSix(ctx, "child", parentOpt)
	if err != nil {
		return nil, err
	}
	return comp, nil
}

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		_, err := NewComponentSixParent(ctx, "comp6")
		if err != nil {
			return err
		}

		return nil
	})
}

File: tests/integration/aliases/python/retype_parents/step2/__main__.py
# Copyright 2016-2022, Pulumi Corporation.  All rights reserved.

import copy

from pulumi import Alias, ComponentResource, ResourceOptions


class Resource(ComponentResource):
    def __init__(self, name, opts=None):
        super().__init__("my:module:Resource", name, None, opts)

# Scenario #6 - Nested parents changing types
class ComponentSix(ComponentResource):
    def __init__(self, name, opts=None):
        # Add an alias that references the old type of this resource...
        aliases = [
            Alias(type_=f"my:module:ComponentSix-v{i}")
            for i in range(0, 100)
        ]

        # ..and then make the super call with the new type of this resource and the added alias.
        opts_copy = copy.copy(opts)
        opts_copy.aliases = aliases

        super().__init__("my:module:ComponentSix-v100", name, None, opts_copy)
        resource = Resource("otherchild", ResourceOptions(parent=self))

class ComponentSixParent(ComponentResource):
    def __init__(self, name, opts=None):
        # Add an alias that references the old type of this resource...
        aliases = [
            Alias(type_=f"my:module:ComponentSixParent-v{i}")
            for i in range(0, 10)
        ]

        # ..and then make the super call with the new type of this resource and the added alias.
        opts = ResourceOptions(aliases=aliases)

        super().__init__("my:module:ComponentSixParent-v10", name, None, opts)
        child = ComponentSix("child", ResourceOptions(parent=self))

comp6 = ComponentSixParent("comp6")